### PR TITLE
New Data Source: cloudflare_zone

### LINKED ISSUE #163: Add new zone data source
## Summary
Introduce new data source `cloudflare_zone`

# cloudflare_zone

Use this data source to look up a [Zone][1] record.

## Example Usage

```hcl
data "cloudflare_zone" "test" {
  zone = "example.com"
}

resource "cloudflare_spectrum_application" "ssh" {
  zone_id  = "${data.cloudflare_zone.test.id}"
  protocol = "tcp/22"

  dns = {
    type = "CNAME"
    name = "ssh.${data.cloudflare_zone.test.zone}"
  }

  origin_direct = ["tcp://81.120.102.10:23"]
  origin_port   = 22
}
```

## Argument Reference

- `zone` - (Required) The name of the zone to match.

## Attributes Reference

- `id`           - Id of the zone
- `status`       - Zone status. Valid values: active, pending, initializing, moved, deleted, deactivated and read only
- `name_servers` - List of name servers
- `type`         - Zone type. Valid values: full or partial

[1]: https://api.cloudflare.com/#zone-properties

I will start a PR for this feature

### PR DESCRIPTION
**Pull request by ewilde**
This PR introduces a new data source `cloudflare_zone`. Resolves #163 

1. Tested website manually
2. Tested website with `make website-test`
3. Tested new data source with new integration tests
`TestAccCloudflareZone`

## Test outputs

*manual website verification*
image

*make website-test*
```
2018-11-23 14:02:46 URL:http://127.0.0.1:4567/docs/providers/cloudflare/ [30335/30335] -> "index.html.tmp.tmp" [1]
2018-11-23 14:02:46 URL:http://127.0.0.1:4567/robots.txt [44/44] -> "robots.txt.tmp" [1]
2018-11-23 14:02:46 URL:http://127.0.0.1:4567/docs/providers/cloudflare/index.html [30335/30335] -> "index.html.tmp.tmp" [1]
2018-11-23 14:02:46 URL:http://127.0.0.1:4567/docs/providers/cloudflare/d/ip_ranges.html [27990/27990] -> "ip_ranges.html.tmp.tmp" [1]
2018-11-23 14:02:46 URL:http://127.0.0.1:4567/docs/providers/cloudflare/d/zone.html [28422/28422] -> "zone.html.tmp.tmp" [1]
2018-11-23 14:02:46 URL:http://127.0.0.1:4567/docs/providers/cloudflare/r/access_application.html [28527/28527] -> "access_application.html.tmp.tmp" [1]
2018-11-23 14:02:46 URL:http://127.0.0.1:4567/docs/providers/cloudflare/r/access_policy.html [32286/32286] -> "access_policy.html.tmp.tmp" [1]
2018-11-23 14:02:46 URL:http://127.0.0.1:4567/docs/providers/cloudflare/r/access_rule.html [33022/33022] -> "access_rule.html.tmp.tmp" [1]
2018-11-23 14:02:46 URL:http://127.0.0.1:4567/docs/providers/cloudflare/r/account_member.html [28507/28507] -> "account_member.html.tmp.tmp" [1]
2018-11-23 14:02:46 URL:http://127.0.0.1:4567/docs/providers/cloudflare/r/custom_pages.html [29553/29553] -> "custom_pages.html.tmp.tmp" [1]
2018-11-23 14:02:46 URL:http://127.0.0.1:4567/docs/providers/cloudflare/r/filter.html [29701/29701] -> "filter.html.tmp.tmp" [1]
2018-11-23 14:02:46 URL:http://127.0.0.1:4567/docs/providers/cloudflare/r/firewall_rule.html [30863/30863] -> "firewall_rule.html.tmp.tmp" [1]
2018-11-23 14:02:46 URL:http://127.0.0.1:4567/docs/providers/cloudflare/r/load_balancer.html [33804/33804] -> "load_balancer.html.tmp.tmp" [1]
2018-11-23 14:02:46 URL:http://127.0.0.1:4567/docs/providers/cloudflare/r/load_balancer_monitor.html [30962/30962] -> "load_balancer_monitor.html.tmp.tmp" [1]
2018-11-23 14:02:46 URL:http://127.0.0.1:4567/docs/providers/cloudflare/r/load_balancer_pool.html [31848/31848] -> "load_balancer_pool.html.tmp.tmp" [1]
2018-11-23 14:02:46 URL:http://127.0.0.1:4567/docs/providers/cloudflare/r/page_rule.html [34527/34527] -> "page_rule.html.tmp.tmp" [1]
2018-11-23 14:02:46 URL:http://127.0.0.1:4567/docs/providers/cloudflare/r/rate_limit.html [35510/35510] -> "rate_limit.html.tmp.tmp" [1]
2018-11-23 14:02:46 URL:http://127.0.0.1:4567/docs/providers/cloudflare/r/record.html [30206/30206] -> "record.html.tmp.tmp" [1]
2018-11-23 14:02:46 URL:http://127.0.0.1:4567/docs/providers/cloudflare/r/waf_rule.html [28375/28375] -> "waf_rule.html.tmp.tmp" [1]
2018-11-23 14:02:46 URL:http://127.0.0.1:4567/docs/providers/cloudflare/r/worker_route.html [31118/31118] -> "worker_route.html.tmp.tmp" [1]
2018-11-23 14:02:46 URL:http://127.0.0.1:4567/docs/providers/cloudflare/r/worker_script.html [30107/30107] -> "worker_script.html.tmp.tmp" [1]
2018-11-23 14:02:46 URL:http://127.0.0.1:4567/docs/providers/cloudflare/r/zone.html [29450/29450] -> "zone.html.tmp.tmp" [1]
2018-11-23 14:02:46 URL:http://127.0.0.1:4567/docs/providers/cloudflare/r/zone_lockdown.html [30919/30919] -> "zone_lockdown.html.tmp.tmp" [1]
2018-11-23 14:02:46 URL:http://127.0.0.1:4567/docs/providers/cloudflare/r/zone_settings_override.html [37636/37636] -> "zone_settings_override.html.tmp.tmp" [1]
Found no broken links.

FINISHED --2018-11-23 14:02:46--
Total wall clock time: 0.7s
Downloaded: 24 files, 697K in 0.04s (15.4 MB/s)
tf-website-cloudflare-temp
make[1]: Leaving directory '/home/ewilde/data/go/src/github.com/hashicorp/terraform-website'
```
**Integration test**
```
GOROOT=/usr/local/go #gosetup
GOPATH=/home/ewilde/data/go #gosetup
/usr/local/go/bin/go test -c -i -o /tmp/___TestAccCloudflareZone_in_github_com_ewilde_terraform_provider_cloudflare_cloudflare github.com/ewilde/terraform-provider-cloudflare/cloudflare #gosetup
/usr/local/go/bin/go tool test2json -t /tmp/___TestAccCloudflareZone_in_github_com_ewilde_terraform_provider_cloudflare_cloudflare -test.v -test.run ^TestAccCloudflareZone$ #gosetup
=== RUN   TestAccCloudflareZone
--- PASS: TestAccCloudflareZone (1.31s)
PASS

Process finished with exit code 0
```

Signed-off-by: Edward Wilde <ewilde@gmail.com>

### CODE REVIEW

**Comment by ewilde** [comment]
@patryk we could add a different data source that listed zones, my use case is to lookup a single zone and is analogous to https://www.terraform.io/docs/providers/aws/d/route53_zone.html which we use currently in our terraform code, the aws `aws_route53_zone` returns a single zone. This is convenient when you only want one zone, you don't need to use the splat syntax etc.. etc..

We are hopefully moving to cloudflare for DNS and this seemed a natural way to port the code

File: cloudflare/data_source_zone.go [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+	return &schema.Resource{
+		Read: dataSourceCloudflareZoneRead,
+
+		Schema: map[string]*schema.Schema{
+			"zone": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ValidateFunc: validation.StringMatch(regexp.MustCompile("^([a-zA-Z0-9][\\-a-zA-Z0-9]*\\.)+[\\-a-zA-Z0-9]{2,20}$"), ""),
+			},
+			"jump_start": {
```
[the remark this answers is not in the document]
Comment (ewilde) [reply cloudflare/data_source_zone.go]: :+1: I'll remove this attribute

**Comment by ewilde** [comment]
@patryk happy to work on a `cloudlfare_zones` as a separate PR, i can raise an issue and we can discuss design

**Comment by danihodovic** [comment]
Agreed with @ewilde . It would be great to see both data sources for a single zone and a list of zones. My guess is that the majority of use cases will use a single zone.

**Comment by ewilde** [comment]
@danihodovic 👍 https://github.com/terraform-providers/terraform-provider-cloudflare/pull/168

**Comment by jacobbednarz** [comment]
Having a look at the two PRs, I'm unsure if there is any benefit to having two that _could_ do similar things based on conditional functionality. Here is a couple of example of some pseudo Terraform resources:

- Get a single resource based on the zone.

  ```hcl
  data "cloudflare_zone" "test" {
    zone = "example.com"
  }
  resource "cloudflare_spectrum_application" "ssh" {
    zone_id  = "${data.cloudflare_zone.test.id}"
    protocol = "tcp/22"
    dns = {
      type = "CNAME"
      name = "ssh.${data.cloudflare_zone.test.zone}"
    }
    origin_direct = ["tcp://81.120.102.10:23"]
    origin_port   = 22
  }
  ```

- Get multiple resources where the DNS label contains "test".

  ```hcl
  data "cloudflare_zone" "test" {
    filter {
      field = "label"
      operator = "contains"
      value = "test"
    }
  }

  resource "cloudflare_spectrum_application" "ssh" {
    zone_id  = "${data.cloudflare_zone.test[0].id}"
    protocol = "tcp/22"
    dns = {
      type = "CNAME"
      name = "ssh.${data.cloudflare_zone.test.zone}"
    }
    origin_direct = ["tcp://81.120.102.10:23"]
    origin_port   = 22
  }
  ```

The logic here would read something like if the `name` attribute was present, do a straight lookup otherwise, rely on passing through the `filter` parameters and sorting on the client side. Ideally the filtering should be done on Cloudflare's end before passing back the results but seeing how the [API](https://api.cloudflare.com/#zone-list-zones) is somewhat limited to the ordering/status/match type, we might need to handroll some of the other operations that we might be missing (unsure of what fields we do want here).

**Comment by patryk** [comment]
> Having a look at the two PRs, I'm unsure if there is any benefit to having two that _could_ do similar things based on conditional functionality.

My thought exactly, I don't see convincing reason having two separate data sources doing almost exact thing, especially `cloudflare_zones` could be used to return just a single zone.

> The logic here would read something like if the `name` attribute was present, do a straight lookup otherwise, rely on passing through the `filter` parameters and sorting on the client side. Ideally the filtering should be done on Cloudflare's end before passing back the results but seeing how the [API](https://api.cloudflare.com/#zone-list-zones) is somewhat limited to the ordering/status/match type, we might need to handroll some of the other operations that we might be missing (unsure of what fields we do want here).

If there is anything limiting in the API, let me know and we will extend it.

**Comment by patryk** [comment]
Thanks @ewilde for contribution, but I'd like to focus on delivering value for `cloudflare_zones` data source, including the use case here.

**Comment by ewilde** [comment]
@patryk I would appreciate a reply to my [previous comment](https://github.com/terraform-providers/terraform-provider-cloudflare/pull/167#issuecomment-442955824). As i'm sure you can appreciate I spent considerable effort working and commenting on this PR

**Comment by patryk** [comment]
> @patryk @jacobbednarz I've always used the `aws` terraform provider as inspiration for organising and designing resources and data sources.
> 
> The pattern I followed is using a data source with a singular resource name i.e. `foo` to return a single `foo` id and a separate data source with the pluralised resource name i.e. `foos` that returns arrays of `fo` id's
> 
> This pattern is consistent with the aws provider, examples of which:
> 
>     * [aws_ami](https://www.terraform.io/docs/providers/aws/d/ami.html)
> 
>     * [aws_amis](https://www.terraform.io/docs/providers/aws/d/amis.html)
> 
>     * [aws_availability_zone](https://www.terraform.io/docs/providers/aws/d/availability_zone.html)
> 
>     * [aws_availability_zones](https://www.terraform.io/docs/providers/aws/d/availability_zones.html)
> 
>     * [aws_ebs_snapshot](https://www.terraform.io/docs/providers/aws/d/ebs_snapshot.html)
> 
>     * [aws_ebs_snapshots](https://www.terraform.io/docs/providers/aws/d/ebs_snapshots.html)
> 
>     * [aws_instance](https://www.terraform.io/docs/providers/aws/d/instance.html)
> 
>     * [aws_instances](https://www.terraform.io/docs/providers/aws/d/instances.html)
> 
> 
> etc....
> 
> What do you think ☝️ ?

I believe what AWS did is unnecessary code duplication. The same value can be delivered by plural versions of said data sources. I love the idea of being able to query zones on an account and I would really want to have #168 merged. I do appreciate the effort you put into both PR, but I think we should focus on #168.

**Comment by jacobbednarz** [comment]
I tend to agree with @patryk on this one and I think there are some things that the AWS provider got wrong early on and they haven't been able to fully deprecate. This provider is far from perfect however I also don't generally copy patterns from other providers as their APIs and workflow aren't the same which can cause some janky work arounds to match functionality.

One example of where this would end up with some issues is if we took `cloudflare_zone` and tried to find a single resource based on the name. The `name` attribute isn't unique within Cloudflare's internal systems which means even if you were after an exact match and provided it as `name = "example.com"` there is a chance there would be multiples returned. Majority of the time, this would be 👌 however if the end user took the resource literally and never anticipated multiples coming back, there would be some very unexpected behaviour. This then leads me to the code duplication issue that @patryk mentions as once we start catering for multiple resources coming back in both `cloudflare_zone` and `cloudflare_zones` we're probably better off with just using the latter with great documentation for both scenarios and not adding `cloudflare_zone`.

For this reason, I think providing the `cloudflare_zones` resource is more suitable (as multiples are always possible) and we put the onus on the end user to do their due diligence on expecting the potential of multiple records.

**Comment by ewilde** [comment]
@jacobbednarz when you implement a single data source look up it's sensible not to allow multiple resources see: https://github.com/terraform-providers/terraform-provider-cloudflare/pull/167/files#diff-603406a9cad328fe22a86b0d1b974894R76

**Comment by ewilde** [comment]
I'm happy to agree to disagree :smile: you guys are the maintainers after all.

As an end-user of terraform providers, I enjoy the simplicity of singe data sources and not having to element look ups. 

For an end-user which is better:

1.  `${cloudflare_zones.zone.id}` 

or 

2.`${element(cloudflare_zones.zone.*.id, count.index)}`

**Pull request closed without merging**